packages/webr/src/outputconnection.c

```diff
@@ -0,0 +1,101 @@
+#include <R.h>
+#include <Rinternals.h>
+#include <R_ext/Connections.h>
+
+# define BUFSIZE 10000
+struct output_con_data {
+  SEXP output;
+  char buf[BUFSIZE];
+  char *line, *cur;
+};
+
+static Rboolean output_open(Rconnection con) {
+  con->isopen = TRUE;
+  return TRUE;
+}
+
+static void output_close(Rconnection con) {
+  con->isopen = FALSE;
+}
+
+static void output_destroy(Rconnection con) {
+  if(con->private) free(con->private);
+}
+
+static int output_vfprintf(Rconnection con, const char *format, va_list ap) {
+  struct output_con_data* data = con->private;
+  int res;
+  char *p, *q;
+  size_t curlen = data->cur - data->buf;
+
+  va_list aq;
+  va_copy(aq, ap);
+  res = vsnprintf(data->cur, BUFSIZE - curlen, format, aq);
+  va_end(aq);
+
+  if (res < 0) {
+    data->cur[0] = '\0';
+  } else if(res >= BUFSIZE - curlen) {
+	  warning("printing of extremely long output is truncated");
+    data->buf[BUFSIZE-1] = '\0';
+	  res = BUFSIZE - curlen - 1;
+  }
+  data->cur += res < 0 ? 0 : res;
+
+  // Check for newlines in buffer and copy completed lines to output
+  for (p = data->line; p < data->cur; p++) {
+    if (*p == '\n') {
+      *p = '\0';
+      const char *names[] = {"type", "data"};
```

Comment:
Needs a `""` sentinel.

```suggestion
      const char *names[] = { "type", "data", "" };
```